src/base/associative_cache.hh

```diff
@@ -0,0 +1,308 @@
+     *   of sets can be calculated dividing this balue by the 'assoc' value
+     * @param associativity number of elements in each associative set
+     * @param entry_size Size of each entry
+     * @param repl_policy replacement policy
+     * @param indexing_policy indexing policy
+     */
+    AssociativeCache(const char *name, const size_t num_entries,
+                     const size_t associativity_,
+                     BaseReplacementPolicy *repl_policy,
+                     BaseIndexingPolicy *indexing_policy,
+                     Entry const &init_val = Entry())
+        : Named(std::string(name)), numEntries(num_entries),
+          associativity(associativity_),
+          replPolicy(repl_policy),
+          indexingPolicy(indexing_policy),
+          entries(num_entries, init_val)
+    {
+        initParams(numEntries, associativity);
+    }
+
+    ~AssociativeCache() = default;
+
+    /**
+     * Clear the entries in the cache.
+     */
+    void clear() {
+        for (auto &entry : entries) {
+            invalidate(&entry);
+        }
+    }
+
+    void init(const size_t num_entries,
+              const size_t associativity_,
+              BaseReplacementPolicy *_repl_policy,
+              BaseIndexingPolicy *_indexing_policy,
+              Entry const &init_val = Entry())
+    {
+        numEntries           = num_entries;
+        associativity        = associativity_;
+        replPolicy           = _repl_policy;
+        indexingPolicy       = _indexing_policy;
+        entries.resize(num_entries, init_val);
+
+        initParams(numEntries, associativity);
+    }
+
+    /**
+     * Get the tag for the addr
+     * @param addr Addr to get the tag for
+     * @return Tag for the address
+     */
+    virtual Addr getTag(const Addr addr) const
+    {
+        return indexingPolicy->extractTag(addr);
+    }
+
+    /**
+     * Do an access to the entry if it exists.
+     * This is required to update the replacement information data.
+     * @param addr key to the entry
+     * @return The entry if it exists
+     */
+    virtual Entry* accessEntry(const Addr addr)
+    {
+        auto entry = findEntry(addr);
+
+        if (entry) {
+            replPolicy->touch(entry->replacementData);
+        }
+
+        return entry;
+    }
+
+    /**
+     * Update the replacement information for an entry
+     * @param Entry to access and upate
+     */
+    virtual void accessEntry(Entry *entry)
+    {
+        replPolicy->touch(entry->replacementData);
+    }
+
+    /**
+     * Find an entry within the set
+     * @param addr key element
+     * @return returns a pointer to the wanted entry or nullptr if it does not
+     *  exist.
+     */
+    virtual Entry* findEntry(const Addr addr) const
+    {
+        auto tag   = getTag(addr);
+
+        auto entries = indexingPolicy->getPossibleEntries(addr);
+
+        for (auto it = entries.begin(); it != entries.end(); it++) {
+            Entry *entry = static_cast<Entry*>(*it);
```

Comment:
I *think* you can replace lines 194-197 by something more concise:
        for (auto &entry : entries) {
            if (entry.matchTag(tag)) {
                return &entry;